### PR TITLE
[IMAGING-258] Prevent exception in TIFF when reading EXIF directory

### PR DESCRIPTION
**Pull request by gwlucastrig**
This change addresses JIRA Issue 258 in which files containing EXIF sub-directories caused the TiffReader to throw an ImageReadException.  The library now reads the directory  successfully.

Unfortunately, I do not have a test file to contribute at this time.   The one I've used for testing is the public-domain "Natural Earth Data"  medium-size Cross-blended Hypsometric Tints file at https://www.naturalearthdata.com/downloads/10m-raster-data/10m-cross-blend-hypso/    Unfortunately, this file is just too large to bundle with the commons-imaging software distribution.

I will see if I can produce a test file sometime in the next few days.

Also, an item off topic...  I was wondering when you thought you might be posting the Alpha 2 release of the library to maven central.  I have been working on a wiki article that describes the new features and would like to coordinate posting it.

### CODE REVIEW

**Comment by kinow** [comment]
Looks good!

>Unfortunately, I do not have a test file to contribute at this time. The one I've used for testing is the public-domain "Natural Earth Data" medium-size Cross-blended Hypsometric Tints file at https://www.naturalearthdata.com/downloads/10m-raster-data/10m-cross-blend-hypso/ Unfortunately, this file is just too large to bundle with the commons-imaging software distribution.

Couldn't we resize/trim the data of this file?

>I will see if I can produce a test file sometime in the next few days.

👍 

>Also, an item off topic... I was wondering when you thought you might be posting the Alpha 2 release of the library to maven central. I have been working on a wiki article that describes the new features and would like to coordinate posting it.

Was planning on doing that a few weeks ago when I had a whole week off, but ended up doing other chores that had piled up during our lock down in New Zealand.

In theory we can do that any time now, I just need a few hours in a weekend to sync repositories, and follow the release process to start the vote in the ASF mailing list. Would you like to finish this PR so that it's included in alpha2 too? Otherwise we can aim at this weekend, or the next one?

Thanks @gwlucastrig !

**Comment by gwlucastrig** [comment]
I am undecided as to whether we should delay the PR until I can produce a sample file.  I've been looking at the Commons Imaging API for writing TIFF files and I haven't figured out how to write this kind of data.  A TIFF file is organized into one or more Image File Directory (IFD) elements, each of which can contain a separate image and/or bundle of metadata. The Commons Imaging API supports writing multiple images. The problem I am dealing with here is that the EXIF information in the sample file I used is actually a special kind of element known as a "subdirectory" which is stored within a larger directory. I haven't been able to figure out whether the Commons Imaging API allows a program to write data in that form. If it doesn't, then making it do so would be a much bigger PR than simply reading it when it's there.

Give me a few days to look into this and I will let you know if it is feasible to create a sample image without a major level of effort.

I suppose that the alternate approach would be to install a version of GDAL (or maybe GIMP) and use it to create a file.  Doing that has a learning curve of its own.

**Comment by gwlucastrig** [comment]
I was able to create a test file that exhibits this problem.  I extended the TIffReadTest unit test to exercise the relevant methods.

I believe that, pending successful checks and reviews, this change closes out JIRA Issue 258.

File: src/test/java/org/apache/commons/imaging/formats/tiff/TiffReadTest.java [hunk 1]
```diff
@@ -1,54 +1,70 @@
...
+			TiffReader tiffReader = new TiffReader(true);
+			TiffContents contents = tiffReader.readDirectories(
+				byteSource,
+				true,
+				FormatCompliance.getDefault());
+				assertNotNull(contents);
+        }
+	}

+}
```
Comment (kinow) [inline src/test/java/org/apache/commons/imaging/formats/tiff/TiffReadTest.java:70]: Missing extraline, and I think it's using Windows CRLF for line breaks?

Comment (gwlucastrig) [reply src/test/java/org/apache/commons/imaging/formats/tiff/TiffReadTest.java]: Yes.  Windows strikes again.  Fix submitted.

**Comment by gwlucastrig** [comment]
Just checking in.  Is there anything else I need to do with this pull request?

**Comment by kinow** [comment]
Ops, not at all @gwlucastrig , just got distracted with other projects. Let's review/merge and we should be ready for alpha2 :-) I can probably review it either tomorrow, or until Sunday (NZ time).

**Comment by gwlucastrig** [comment]
Bruno,

Thanks.  Don't worry about rushing the Alpha-2 release just for me.  Please do what's best for the project.  I'm mainly concerned about getting the Pull Request closed out because I am looking into contributing more enhancements for Commons Imaging and haven't really mastered the art of managing two PR's at once (I must have some kind of mental block in terms of git, but I'll figure it out eventually).

My sister used to live in New Zealand and loved the place.  She's always encouraging me to visit.  Maybe after this pandemic is over.

File: src/main/java/org/apache/commons/imaging/formats/tiff/taginfos/TagInfoDirectory.java [hunk 2]
```diff
@@ -16,14 +16,23 @@
...
+import java.util.List;
 import org.apache.commons.imaging.formats.tiff.constants.TiffDirectoryType;
+import org.apache.commons.imaging.formats.tiff.fieldtypes.FieldType;
 
 /**
  * A LONG representing an offset to a TIFF directory.
  */
 public class TagInfoDirectory extends TagInfoLong {
+
+    static final List<FieldType> fieldList
```
Comment (kinow) [inline src/main/java/org/apache/commons/imaging/formats/tiff/taginfos/TagInfoDirectory.java:30]: Let's leave this one `private` until we need to make it public? It appears to be used only on this class.

**Comment by kinow** [comment]
I've also updated the issue title to include the JIRA issue. Just to confirm, this PR is fixing IMAGING-258, right?

**Comment by gwlucastrig** [comment]
Yes.  This change is specifically for Imaging 258.

**Comment by kinow** [comment]
Merged, thanks @gwlucastrig !

**Comment by kinow** [comment]
@gwlucastrig Commons Imaging 1.0-alpha2 release vote thread: https://markmail.org/thread/vkmhtk6xo4t5ncf7

That is open for 72 hours. If there are no blockers, the release should be completed by the middle of the next week, New Zealand time.

Lots of improvements you have contributed to are included in this release :medal_sports: 

Thanks a lot!
Bruno

**Pull request closed without merging**